uaclient/entitlements/livepatch.py

```diff
@@ -26,9 +38,235 @@
+        query_params = {
+            "kernel-version": version,
+            "flavour": flavor,
+            "architecture": arch,
+            "codename": codename,
+        }
+        headers = self.headers()
+        try:
+            result, _headers = self.request_url(
+                LIVEPATCH_API_V1_KERNELS_SUPPORTED,
+                query_params=query_params,
+                headers=headers,
+            )
+        except Exception as e:
+            with util.disable_log_to_console():
+                logging.warning(
+                    "error while checking livepatch supported kernels API"
+                )
+                logging.warning(e)
+            return None
+
+        if not isinstance(result, dict):
+            return None
+
+        supported = result.get("Supported", None)
+        if supported == "yes":
+            return True
+
+        return False
+
+
+class LivepatchSupportCacheData(DataObject):
+    fields = [
+        Field("version", StringDataValue),
+        Field("flavor", StringDataValue),
+        Field("arch", StringDataValue),
+        Field("codename", StringDataValue),
+        Field("supported", BoolDataValue, required=False),
+        Field("cached_at", DatetimeDataValue),
+    ]
+
+    def __init__(
+        self,
+        version: str,
+        flavor: str,
+        arch: str,
+        codename: str,
+        supported: Optional[bool],
+        cached_at: datetime.datetime,
+    ):
+        self.version = version
+        self.flavor = flavor
+        self.arch = arch
+        self.codename = codename
+        self.supported = supported
+        self.cached_at = cached_at
+
+
+LIVEPATCH_SUPPORT_CACHE = files.DataObjectFile(
+    LivepatchSupportCacheData,
+    files.UAFile(
+        "livepatch-kernel-support-cache.json",
+        "/run/ubuntu-advantage",
+        private=False,
+    ),
+    file_format=files.DataObjectFileFormat.JSON,
+)
+
+
+def _on_supported_kernel_cli() -> Optional[bool]:
+    if is_livepatch_installed():
+        try:
+            out, _ = system.subp([LIVEPATCH_CMD, "status", "--format", "json"])
+        except exceptions.ProcessExecutionError:
+            logging.warning(
+                "canonical-livepatch returned error when "
+                "checking kernel support"
+            )
+            return None
+
+        cli_statuses = json.loads(out).get("Status", [])
+        if len(cli_statuses) > 0:
+            cli_status = cli_statuses[0].get("Livepatch", {})
+
+            cli_supported = cli_status.get("Supported", None)
+            if cli_supported == "supported":
+                return True
+            if cli_supported == "unsupported":
+                return False
+
+            cli_state = cli_status.get("State", None)
+            if cli_state == "applied":
+                logging.debug(
+                    "kernel must be supported because livepatches "
+                    " have been applied"
+                )
+                return True
```

Comment:
We actually may not be able to assume this means the kernel is supported. For example, if a kernel is supported for a while but then we end support, then we need to warn the user that their support has ended even if the kernel has patches currently applied.